Fix unordered posting to alt-da server in case of concurrency > 1

nktaushanov opened the pull request:
<!--
Contributions welcome! See https://github.com/ethereum-optimism/.github/blob/master/CONTRIBUTING.md
-->

**Description**

When using `--altda.max-concurrent-da-request=1` or not set at all, all batches are posted to the alt-da server sequently and the server responds, the commitments are added to a queue for posting to L1. But if we set it to more than 1, the batcher sends requests to the alt-da server and then adds to the same queue for posting to the l1 in the order of the server's responses. This means that depending on when responses will get processed, the batcher can post to L1 unordered commitments. Since Holocene, this is not supported and op-node will drop the batch during derivation, causing the batcher to try and post again.

Current fix introduces a new go routine which is only responsible for posting to the L1. Alt-da server responses are posted to a channel and then processed by this routine in the order that they were first send. This ensures that we will post commitments in orderly fashion and op-node will derive successfully.

This change is minimal and aims at only touching places relevant to the alt-da logic.

**Tests**

There is a basic unit test covering the new functionality. I did also load tested the batcher with concurrency=12 to make sure things won't break or hang.

nktaushanov commented:
@geoknee Mind enabling CI on this PR? Thank you!

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates

pcw109550 commented:
/ci authorize https://github.com/ethereum-optimism/optimism/pull/18478/changes/71eaa75fcb91e6fec7f7cb28043ca02d068eee1f

geoknee commented:
/ci authorize [50466e4](https://github.com/ethereum-optimism/optimism/pull/18478/commits/50466e4f42fb07315a9b229d532ce3b84e2dd9af)

geoknee commented on op-alt-da/cli.go:
```diff
@@ -99,6 +99,10 @@ func (c CLIConfig) Check() error {
 			return fmt.Errorf("DA server URL is invalid: %w", err)
 		}
+
+		if c.MaxConcurrentRequests < 1 {
+			return fmt.Errorf("max concurrent requests must be at least 1")
+		}
```
Can you explain this change?

nktaushanov replied:
MaxConcurrentRequests is 0 would implicitly mean unbound concurrency. But then current implementation uses this number to define the commitmentsCh and 0 there means blocking buffered channel and as there are no unbuffered channels in golang, implementing this "unbound concurrency" adds a lot of complexity which does not seem practical. Processing the commitments is bound by another flag which is the number of pending l1 transactions. So for the sake of keeping things simpler, I decided to remove this case. Even default here is 1. Let me know if you think 0 is required and we can try to brainstorm how it should work in combination of the rest of the flags.

geoknee replied:
I don't think 0 is required. However, I think this new condition is breaking some tests in `op-alt-da`. PTAL.

nktaushanov replied:
addressed these, should pass now

nktaushanov replied:
addressed these, should pass now

geoknee commented:
The fix makes sense to me, I just want to check with the rest of the team about organising all channels and getting crisp about whether we pass them as arguments or access them through state. Ideally the altDA code is very well isolated from the rest of the batcher, not sure if it is possible to do that here or not.

nktaushanov commented:
> The fix makes sense to me, I just want to check with the rest of the team about organising all channels and getting crisp about whether we pass them as arguments or access them through state. Ideally the altDA code is very well isolated from the rest of the batcher, not sure if it is possible to do that here or not.

I do agree with you and such change makes a lot of sense, that was my first thought as well. Trying to implement it though lead to a huge change in the batcher. Given we are trying to fix a bug here, I decided it's better to keep it simple and change the minimal amount of work and if you think the logic should be refactored, we can do that in a follow up/another change which only focuses on that. The combination of both makes it much harder to read and prove the behavior is preserved and fixed in the same time

nktaushanov committed:
fix(op-batcher): preserve alt-DA commitment ordering with concurrent requests

Introduce a commitmentPayloadChan to ensure alt-DA commitments are
posted to L1 in the same order they were submitted, even when DA
requests complete out of order due to concurrency. Also validate
MaxConcurrentRequests >= 1 in CLIConfig.Check().

geoknee commented:
/ci authorize [f059f97](https://github.com/ethereum-optimism/optimism/commit/f059f974ab56c08b7b402405e4491da4223321e6)

nktaushanov committed:
Merge branch 'develop' into op-batcher-alt-da-concurrency-2

nktaushanov commented:
Hey @geoknee, I think we should time this better, I had to resolve another conflict now and you would have to trigger the ci again

geoknee commented:
/ci authorize [989a647](https://github.com/ethereum-optimism/optimism/pull/18478/commits/989a647465457ebff8ae4e9dcf3563de52f23235)

nktaushanov committed:
Fix import

nktaushanov commented:
There was an issue from the merge conflict I did by hand in github, should be good now

geoknee commented:
/ci authorize [532feac](https://github.com/ethereum-optimism/optimism/pull/18478/commits/532feac66b3755a85618b6f08a01a90a94c92bae)

nktaushanov commented:
@geoknee should we merge this?

geoknee commented on op-batcher/batcher/driver.go:
```diff
@@ -953,27 +972,41 @@ func (l *BatchSubmitter) publishToAltDAAndL1(txdata txData, queue *txmgr.Queue[t
+
+func (l *BatchSubmitter) publishAltDACommitmentsToL1Loop(done chan struct{}, queue *txmgr.Queue[txRef], receiptsCh chan txmgr.TxReceipt[txRef], commitmentsCh chan commitmentPayloadChan) {
+	defer close(done)
+	for commitmentCh := range commitmentsCh {
+		for commData := range commitmentCh {
```
This is a little eye-catching. Can't we simplify this, and just use a simple `chan commitmentPayload` instead of a channel of channels?

AFAICT the producer buffers, sends on and closes each inner channel sequentially and then sends it on the outer channel. I think it could just send directly on the outer channel?

geoknee replied:
Ok I think I get it. You send the buffered inner channel on the outer channel synchronously, ensuring that those sends are received in order, irrespective of when the goroutine which is actually going to send the txdata does its send. So we have essentially a channel of "promise"-like objects (channels with a buffer length of 1). 

Assuming that understanding is correct, could we add some code comments to spell this out for future readers of the code? 

Also, something like this on the inner channel could be clearer:

```go
    if commData, ok := <-commitmentCh; ok {
        l.Log.Info("Processing and sending to l1", "commitment", commData.comm)
        candidate := l.calldataTxCandidate(commData.comm.TxData())
        l.sendTx(commData.txdata, false, candidate, queue, receiptsCh)
    }
```
we don't need to range over tha channel if we know it is only going to send one value.

geoknee replied:
The rest of the changes LGTM, if we can do this bit of tidy up then we get this merged.

kien-rise replied:
Hi, I’m also looking into this issue.

One concern I have is that this PR has not handled DA failures properly. More specifically, `commitmentCh` can be empty (e.g., due to an error when uploading to DA). In that case, Holocene’s strict ordering is no longer guaranteed.

---

**Concrete example**

1. Suppose the Batcher is submitting channels 101, 102, and 103.
2. The `txdata` for 101, 102, and 103 are sent to DA.
3. After a few seconds, DA responds: 101 = ok, 102 = fail, 103 = ok.
4. Since 102 fails during the DA upload, it is marked as failed and scheduled for retry.
5. `publishAltDACommitmentsToL1Loop` runs and submits the L1 transactions for 101 and 103. There is no commitment for 102 at this point.
6. Channel 102 is retried. This time, the DA upload succeeds and the L1 transaction also succeeds.
7. The Batcher now considers everything successful and proceeds with 104, 105, 106, etc.

From the `op-node`’s perspective, the observed order is: 101, 103, 102 — which is incorrect. In fact, 103 will be rejected due to the invalid order, while 101 and 102 are accepted. From that point onward, `op-node` waits for 103, but it will never arrive.

---

I’m currently exploring possible solutions, but it’s still unclear what the best approach is.



<details>
<summary>As a separate attempt to enforce Holocene’s strict ordering, I implemented the following. It’s not very elegant, but I’m sharing it here for discussion purposes.</summary>

---

```go
// isLikelyHolocene returns true if Holocene is likely active based on the current system clock.
func (s *channelManager) isLikelyHolocene() bool {
	return s.rollupCfg.IsHolocene(uint64(time.Now().Unix()))
}

// TxFailed records a transaction as failed. It will attempt to resubmit the data
// in the failed transaction.
func (s *channelManager) TxFailed(_id txID) {
	id := _id.String()
	if channel := s.getChannel(_id.ChannelID()); channel != nil {
		channel.TxFailed(id)

		if s.isLikelyHolocene() {
			// Under Holocene, strict ordering is required: all the transactions after the failed
			// one in the queue must be submitted again.
			if idx := slices.Index(s.channelQueue, channel); idx >= 0 {
				for _, ch := range s.channelQueue[idx+1:] {
					if ch.frameCursor > 0 {
						s.log.Warn("Holocene active: resetting frame cursor of subsequent channel due to failed tx",
							"channel_id", ch.ID(),
							"failed_tx", id)
						ch.frameCursor = 0
					}
				}
			}
		}
	} else {
		s.log.Warn("transaction from unknown channel marked as failed", "id", id)
	}
}

// TxConfirmed marks a transaction as confirmed on L1. Only if the channel timed out
// the channelManager's state is modified.
func (s *channelManager) TxConfirmed(_id txID, inclusionBlock eth.BlockID) {
	id := _id.String()
	if channel := s.getChannel(_id.ChannelID()); channel != nil {
		if s.isLikelyHolocene() && _id.FirstFrameNumber() >= uint16(channel.frameCursor) {
			// Under Holocene, frames must be submitted in strict order. If the frame cursor has been
			// rewound past this tx's frames, they were already resubmitted and this confirmation is stale.
			s.log.Warn("tx confirmed out of order, frames will be resubmitted", "channel_id", channel.ID(), "tx_id", id)
		} else if timedOut := channel.TxConfirmed(id, inclusionBlock); timedOut {
			s.log.Warn("channel timed out on chain", "channel_id", channel.ID(), "tx_id", id)
			s.handleChannelInvalidated(channel)
		}
	} else {
		s.log.Warn("transaction from unknown channel marked as confirmed", "id", id)
	}
	s.metr.RecordBatchTxSubmitted()
	s.log.Debug("marked transaction as confirmed", "id", id, "block", inclusionBlock)
}
```

---

</details>

kien-rise replied:
Eventually, I chose a simple solution, which is to add a while loop in the goroutine:

```go
for attempt := 1; comm == nil; attempt++ {
	comm, err = l.AltDA.SetInput(l.shutdownCtx, txdata.CallData())
	if err == nil {
		break
	}
	if errors.Is(err, context.Canceled) {
		return nil
	} else {
		... // sleep for a while
	}
}
```

opgitgovernance commented:
This pr has been automatically marked as stale and will be closed in 5 days if no updates
